File: code/packages/fsharp/directed-graph/DirectedGraph.fs
namespace CodingAdventures.DirectedGraph.FSharp

open System
open System.Collections.Generic

type CycleError(message: string, cycle: IReadOnlyList<string>) =
    inherit Exception(message)

    member _.Cycle = cycle

type NodeNotFoundError(node: string) =
    inherit Exception(sprintf "Node not found: \"%s\"" node)

    member _.Node = node

type EdgeNotFoundError(fromNode: string, toNode: string) =
    inherit Exception(sprintf "Edge not found: \"%s\" -> \"%s\"" fromNode toNode)

    member _.FromNode = fromNode
    member _.ToNode = toNode

type Graph(?allowSelfLoops: bool) =
    let allowSelfLoops = defaultArg allowSelfLoops false
    let forward = Dictionary<string, HashSet<string>>(StringComparer.Ordinal)
    let reverse = Dictionary<string, HashSet<string>>(StringComparer.Ordinal)

    let ensureNode (node: string) =
        if not (forward.ContainsKey(node)) then
            raise (NodeNotFoundError(node))

    let orderedStrings (values: seq<string>) =
        values |> Seq.sortWith (fun left right -> StringComparer.Ordinal.Compare(left, right)) |> Seq.toList

    member _.AllowSelfLoops = allowSelfLoops
    member _.Size = forward.Count

    member _.AddNode(node: string) =
        if not (forward.ContainsKey(node)) then
            forward.[node] <- HashSet<string>(StringComparer.Ordinal)

        if not (reverse.ContainsKey(node)) then
            reverse.[node] <- HashSet<string>(StringComparer.Ordinal)

    member this.RemoveNode(node: string) =
        ensureNode node

        for successor in forward.[node] |> Seq.toArray do
            reverse.[successor].Remove(node) |> ignore

        for predecessor in reverse.[node] |> Seq.toArray do
            forward.[predecessor].Remove(node) |> ignore

        forward.Remove(node) |> ignore
        reverse.Remove(node) |> ignore

    member _.HasNode(node: string) = forward.ContainsKey(node)

    member _.Nodes() = orderedStrings forward.Keys

    member this.AddEdge(fromNode: string, toNode: string) =
        if fromNode = toNode && not allowSelfLoops then
            invalidArg "toNode" (sprintf "Self-loops are not allowed: \"%s\" -> \"%s\"" fromNode toNode)

        this.AddNode(fromNode)
        this.AddNode(toNode)
        forward.[fromNode].Add(toNode) |> ignore
        reverse.[toNode].Add(fromNode) |> ignore

    member _.RemoveEdge(fromNode: string, toNode: string) =
        ensureNode fromNode

        if not (forward.ContainsKey(toNode)) || not (forward.[fromNode].Contains(toNode)) then
            raise (EdgeNotFoundError(fromNode, toNode))

        forward.[fromNode].Remove(toNode) |> ignore
        reverse.[toNode].Remove(fromNode) |> ignore

    member _.HasEdge(fromNode: string, toNode: string) =
        forward.ContainsKey(fromNode) && forward.[fromNode].Contains(toNode)

    member _.Successors(node: string) =
        ensureNode node
        orderedStrings forward.[node]

    member _.Predecessors(node: string) =
        ensureNode node
        orderedStrings reverse.[node]

    member _.Edges() =
        [
            for KeyValue(source, targets) in forward do
                for target in targets do
                    yield source, target
        ]
        |> List.sortWith (fun (leftSource, leftTarget) (rightSource, rightTarget) ->
            let bySource = StringComparer.Ordinal.Compare(leftSource, rightSource)
            if bySource <> 0 then bySource else StringComparer.Ordinal.Compare(leftTarget, rightTarget))

    member this.TransitiveClosure(startNode: string) =
        this.Reach(startNode, this.Successors)

    member this.TransitiveDependents(startNode: string) =
        this.Reach(startNode, this.Predecessors)

    member this.TopologicalSort() =
        let indegree = Dictionary<string, int>(StringComparer.Ordinal)
        for node in forward.Keys do
            indegree.[node] <- reverse.[node].Count

        let available =
            ResizeArray(
                indegree
                |> Seq.choose (fun (KeyValue(node, count)) -> if count = 0 then Some node else None)
                |> Seq.sortWith (fun left right -> StringComparer.Ordinal.Compare(left, right)))
        let order = ResizeArray<string>()

        let popSmallest () =
            let smallest =
                available
                |> Seq.minBy id

            available.Remove(smallest) |> ignore
            smallest

        while available.Count > 0 do
            let node = popSmallest ()
            order.Add(node)

            for successor in this.Successors(node) do
                indegree.[successor] <- indegree.[successor] - 1
                if indegree.[successor] = 0 then
                    available.Add(successor)

        if order.Count <> forward.Count then
            raise (CycleError("Graph contains a cycle", this.FindCycle()))

        order |> Seq.toList

    member this.IndependentGroups() =
        let indegree = Dictionary<string, int>(StringComparer.Ordinal)
        for node in forward.Keys do
            indegree.[node] <- reverse.[node].Count

        let remaining = HashSet<string>(forward.Keys, StringComparer.Ordinal)
        let groups = ResizeArray<IReadOnlyList<string>>()

        while remaining.Count > 0 do
            let layer =
                remaining
                |> Seq.filter (fun node -> indegree.[node] = 0)
                |> Seq.sortWith (fun left right -> StringComparer.Ordinal.Compare(left, right))
                |> Seq.toList

            if List.isEmpty layer then
                raise (CycleError("Graph contains a cycle", this.FindCycle()))

            groups.Add(layer)
            for node in layer do
                remaining.Remove(node) |> ignore
                for successor in this.Successors(node) do
                    indegree.[successor] <- indegree.[successor] - 1

        groups |> Seq.toList

    member this.AffectedNodes(changedNodes: seq<string>) =
        let affected = HashSet<string>(StringComparer.Ordinal)
        for node in changedNodes do
            if this.HasNode(node) then
                affected.Add(node) |> ignore
                affected.UnionWith(this.TransitiveClosure(node))

        this.TopologicalSort() |> List.filter affected.Contains

    member private _.Reach(startNode: string, next: string -> string list) =
        ensureNode startNode

        let visited = HashSet<string>(StringComparer.Ordinal)
        let stack = Stack<string>()
        stack.Push(startNode)

        while stack.Count > 0 do
            let node = stack.Pop()
            for neighbor in next(node) do
                if visited.Add(neighbor) then
                    stack.Push(neighbor)

        visited :> ISet<string>

    member private this.FindCycle() =
        let visited = HashSet<string>(StringComparer.Ordinal)
        let onStack = HashSet<string>(StringComparer.Ordinal)
        let parent = Dictionary<string, string option>(StringComparer.Ordinal)

        let rec findCycleFrom node =
            if onStack.Contains(node) then
                let mutable cycle = [ node ]
                let mutable current =
                    match parent.TryGetValue(node) with
                    | true, value -> value
                    | _ -> None

                while current.IsSome && current.Value <> node do
                    cycle <- current.Value :: cycle
                    current <-
                        match parent.TryGetValue(current.Value) with
                        | true, value -> value
                        | _ -> None

                cycle @ [ node ]

            elif not (visited.Add(node)) then
                []

            else
                onStack.Add(node) |> ignore
                let mutable found = []
                for successor in this.Successors(node) do
                    if List.isEmpty found then
                        parent.[successor] <- Some node
                        let cycle = findCycleFrom successor
                        if not (List.isEmpty cycle) then
                            found <- cycle

                onStack.Remove(node) |> ignore
                found

        forward.Keys
        |> Seq.tryPick (fun node ->
            let cycle = findCycleFrom node
            if List.isEmpty cycle then None else Some cycle)
        |> Option.defaultValue []

type LabeledDirectedGraph(?allowSelfLoops: bool) =
    let graph = Graph(defaultArg allowSelfLoops false)
    let labels = Dictionary<string, Dictionary<string, HashSet<string>>>(StringComparer.Ordinal)

    member _.AddNode(node: string) =
        graph.AddNode(node)
        if not (labels.ContainsKey(node)) then
            labels.[node] <- Dictionary<string, HashSet<string>>(StringComparer.Ordinal)

    member this.AddEdge(fromNode: string, toNode: string, label: string) =
        this.AddNode(fromNode)
        this.AddNode(toNode)
        graph.AddEdge(fromNode, toNode)

        if not (labels.[fromNode].ContainsKey(toNode)) then
            labels.[fromNode].[toNode] <- HashSet<string>(StringComparer.Ordinal)

        labels.[fromNode].[toNode].Add(label) |> ignore

    member _.Labels(fromNode: string, toNode: string) =
        match labels.TryGetValue(fromNode) with
        | true, targets when targets.ContainsKey(toNode) ->
            targets.[toNode] |> Seq.sortWith (fun left right -> StringComparer.Ordinal.Compare(left, right)) |> Seq.toList
        | _ ->
            []

    member _.TopologicalSort() = graph.TopologicalSort()
    member _.TransitiveClosure(startNode: string) = graph.TransitiveClosure(startNode)

module DirectedGraph =
    let create allowSelfLoops = Graph(allowSelfLoops)
    let createDefault () = Graph()

    let addEdge fromNode toNode (graph: Graph) =
        graph.AddEdge(fromNode, toNode)
        graph

    let topologicalSort (graph: Graph) = graph.TopologicalSort()

File: code/packages/csharp/directed-graph/DirectedGraph.cs
namespace CodingAdventures.DirectedGraph;

public sealed class CycleError : Exception
{
    public CycleError(string message, IReadOnlyList<string> cycle) : base(message)
    {
        Cycle = cycle;
    }

    public IReadOnlyList<string> Cycle { get; }
}

public sealed class NodeNotFoundError : Exception
{
    public NodeNotFoundError(string node) : base($"Node not found: \"{node}\"")
    {
        Node = node;
    }

    public string Node { get; }
}

public sealed class EdgeNotFoundError : Exception
{
    public EdgeNotFoundError(string fromNode, string toNode) : base($"Edge not found: \"{fromNode}\" -> \"{toNode}\"")
    {
        FromNode = fromNode;
        ToNode = toNode;
    }

    public string FromNode { get; }
    public string ToNode { get; }
}

public sealed class Graph
{
    private readonly Dictionary<string, HashSet<string>> _forward = new(StringComparer.Ordinal);
    private readonly Dictionary<string, HashSet<string>> _reverse = new(StringComparer.Ordinal);

    public Graph(bool allowSelfLoops = false)
    {
        AllowSelfLoops = allowSelfLoops;
    }

    public bool AllowSelfLoops { get; }
    public int Size => _forward.Count;

    public void AddNode(string node)
    {
        _forward.TryAdd(node, new HashSet<string>(StringComparer.Ordinal));
        _reverse.TryAdd(node, new HashSet<string>(StringComparer.Ordinal));
    }

    public void RemoveNode(string node)
    {
        if (!_forward.ContainsKey(node))
        {
            throw new NodeNotFoundError(node);
        }

        foreach (var successor in _forward[node].ToArray())
        {
            _reverse[successor].Remove(node);
        }

        foreach (var predecessor in _reverse[node].ToArray())
        {
            _forward[predecessor].Remove(node);
        }

        _forward.Remove(node);
        _reverse.Remove(node);
    }

    public bool HasNode(string node) => _forward.ContainsKey(node);
    public IReadOnlyList<string> Nodes() => _forward.Keys.ToList();

    public void AddEdge(string fromNode, string toNode)
    {
        if (fromNode == toNode && !AllowSelfLoops)
        {
            throw new ArgumentException($"Self-loops are not allowed: \"{fromNode}\" -> \"{toNode}\"");
        }

        AddNode(fromNode);
        AddNode(toNode);
        _forward[fromNode].Add(toNode);
        _reverse[toNode].Add(fromNode);
    }

    public void RemoveEdge(string fromNode, string toNode)
    {
        if (!HasNode(fromNode))
        {
            throw new NodeNotFoundError(fromNode);
        }

        if (!HasNode(toNode) || !_forward[fromNode].Contains(toNode))
        {
            throw new EdgeNotFoundError(fromNode, toNode);
        }

        _forward[fromNode].Remove(toNode);
        _reverse[toNode].Remove(fromNode);
    }

    public bool HasEdge(string fromNode, string toNode) => HasNode(fromNode) && _forward[fromNode].Contains(toNode);

    public IReadOnlyList<string> Successors(string node)
    {
        if (!HasNode(node))
        {
            throw new NodeNotFoundError(node);
        }

        return _forward[node].ToList();
    }

    public IReadOnlyList<string> Predecessors(string node)
    {
        if (!HasNode(node))
        {
            throw new NodeNotFoundError(node);
        }

        return _reverse[node].ToList();
    }

    public IReadOnlyList<(string From, string To)> Edges() =>
        _forward.SelectMany(pair => pair.Value.Select(target => (pair.Key, target))).ToList();

    public ISet<string> TransitiveClosure(string startNode) => Reach(startNode, Successors);
    public ISet<string> TransitiveDependents(string startNode) => Reach(startNode, Predecessors);

    public IReadOnlyList<string> TopologicalSort()
    {
        var indegree = _forward.Keys.ToDictionary(node => node, node => _reverse[node].Count, StringComparer.Ordinal);
        var queue = new Queue<string>(indegree.Where(pair => pair.Value == 0).Select(pair => pair.Key));
        var order = new List<string>();

        while (queue.Count > 0)
        {
            var node = queue.Dequeue();
            order.Add(node);
            foreach (var successor in _forward[node])
            {
                indegree[successor]--;
                if (indegree[successor] == 0)
                {
                    queue.Enqueue(successor);
                }
            }
        }

        if (order.Count != _forward.Count)
        {
            throw new CycleError("Graph contains a cycle", FindCycle());
        }

        return order;
    }

    public IReadOnlyList<IReadOnlyList<string>> IndependentGroups()
    {
        var indegree = _forward.Keys.ToDictionary(node => node, node => _reverse[node].Count, StringComparer.Ordinal);
        var remaining = new HashSet<string>(_forward.Keys, StringComparer.Ordinal);
        var result = new List<IReadOnlyList<string>>();

        while (remaining.Count > 0)
        {
            var layer = remaining.Where(node => indegree[node] == 0).OrderBy(node => node, StringComparer.Ordinal).ToList();
            if (layer.Count == 0)
            {
                throw new CycleError("Graph contains a cycle", FindCycle());
            }

            result.Add(layer);
            foreach (var node in layer)
            {
                remaining.Remove(node);
                foreach (var successor in _forward[node])
                {
                    indegree[successor]--;
                }
            }
        }

        return result;
    }

    public IReadOnlyList<string> AffectedNodes(IEnumerable<string> changedNodes)
    {
        var affected = new HashSet<string>(StringComparer.Ordinal);
        foreach (var node in changedNodes)
        {
            if (!HasNode(node))
            {
                continue;
            }

            affected.Add(node);
            affected.UnionWith(TransitiveClosure(node));
        }

        return TopologicalSort().Where(affected.Contains).ToList();
    }

    private HashSet<string> Reach(string startNode, Func<string, IReadOnlyList<string>> next)
    {
        if (!HasNode(startNode))
        {
            throw new NodeNotFoundError(startNode);
        }

        var visited = new HashSet<string>(StringComparer.Ordinal);
        var stack = new Stack<string>();
        stack.Push(startNode);
        while (stack.Count > 0)
        {
            var node = stack.Pop();
            foreach (var neighbor in next(node))
            {
                if (visited.Add(neighbor))
                {
                    stack.Push(neighbor);
                }
            }
        }

        return visited;
    }

    private List<string> FindCycle()
    {
        var visited = new HashSet<string>(StringComparer.Ordinal);
        var onStack = new HashSet<string>(StringComparer.Ordinal);
        var parent = new Dictionary<string, string?>(StringComparer.Ordinal);

        foreach (var node in _forward.Keys)
        {
            var cycle = FindCycleFrom(node, visited, onStack, parent);
            if (cycle.Count > 0)
            {
                return cycle;
            }
        }

        return [];
    }

    private List<string> FindCycleFrom(
        string node,
        HashSet<string> visited,
        HashSet<string> onStack,
        Dictionary<string, string?> parent)
    {
        if (onStack.Contains(node))
        {
            var cycle = new List<string> { node };
            var current = parent.GetValueOrDefault(node);
            while (current is not null && current != node)
            {
                cycle.Insert(0, current);
                current = parent.GetValueOrDefault(current);
            }

            cycle.Add(node);
            return cycle;
        }

        if (!visited.Add(node))
        {
            return [];
        }

        onStack.Add(node);
        foreach (var successor in _forward[node])
        {
            parent[successor] = node;
            var cycle = FindCycleFrom(successor, visited, onStack, parent);
            if (cycle.Count > 0)
            {
                return cycle;
            }
        }

        onStack.Remove(node);
        return [];
    }
}

public sealed class LabeledDirectedGraph
{
    private readonly Dictionary<string, Dictionary<string, HashSet<string>>> _labels = new(StringComparer.Ordinal);
    private readonly Graph _graph;

    public LabeledDirectedGraph(bool allowSelfLoops = false)
    {
        _graph = new Graph(allowSelfLoops);
    }

    public void AddNode(string node)
    {
        _graph.AddNode(node);
        _labels.TryAdd(node, new Dictionary<string, HashSet<string>>(StringComparer.Ordinal));
    }

    public void AddEdge(string fromNode, string toNode, string label)
    {
        AddNode(fromNode);
        AddNode(toNode);
        _graph.AddEdge(fromNode, toNode);
        if (!_labels[fromNode].TryGetValue(toNode, out var labels))
        {
            labels = new HashSet<string>(StringComparer.Ordinal);
            _labels[fromNode][toNode] = labels;
        }

        labels.Add(label);
    }

    public IReadOnlyList<string> Labels(string fromNode, string toNode)
    {
        if (_labels.TryGetValue(fromNode, out var targets) && targets.TryGetValue(toNode, out var labels))
        {
            return labels.ToList();
        }

        return [];
    }

    public IReadOnlyList<string> TopologicalSort() => _graph.TopologicalSort();
    public ISet<string> TransitiveClosure(string startNode) => _graph.TransitiveClosure(startNode);
}
